File: src/main/java/org/cip4/xjdf/json/openapi/AttributeType.java
package org.cip4.xjdf.json.openapi;

import org.w3c.dom.Node;

import javax.xml.xpath.XPath;
import java.util.Arrays;
import java.util.List;

public abstract class AttributeType implements Modelable {

    public interface Factory<T extends AttributeType> {
        boolean supports(Node candidate, XPath xPath);

        T create(Node node, Context context);
    }

    static AttributeType create(Node node, Context context) {
        List<Factory<? extends AttributeType>> factories = Arrays.asList(
            new AttributeReferencedType.Factory(),
            new LocalType.Factory()
        );

        return factories.stream()
                        .filter(factory -> factory.supports(node, context.getXPath()))
                        .findFirst()
                        .orElseThrow(() -> new IllegalStateException("Unsupported attribute detected..."))
                        .create(node, context);
    }
}

File: src/main/kotlin/org/cip4/xjdf/openapi/AttributeType.kt
package org.cip4.xjdf.openapi

import org.cip4.xjdf.openapi.model.Schema
import org.w3c.dom.Node
import javax.xml.xpath.XPath

abstract class AttributeType : OpenApiParticle<Schema>() {
    interface Factory<T : AttributeType> {

        fun supports(candidate: Node, xPath: XPath): Boolean

        fun create(node: Node, context: Context): T

        companion object {
            fun create(node: Node, context: Context): AttributeType {
                val factories = listOf(
                    AttributeReferencedType.Factory,
                    LocalType.Factory,
                )
                val factory = factories.find { factory ->
                    factory.supports(node, context.xPath)
                } ?: throw IllegalStateException("Unsupported attribute detected...")
                return factory.create(node, context)
            }
        }

    }
}